### PR TITLE
Sidecar experimental

### PR DESCRIPTION
**Pull request by LimKianAn**
@majst01 @eberlep 
This branch shows the possibility of having the items in `ConfigMap` aligned with *yaml* manifests of target *golang struct* and then unmarshal them in the code so that we avoid setting the values manually. We might not need to merge this branch into *main*, but it serves as an example for further discussion.

### CODE REVIEW

**Comment by LimKianAn** [comment]
Ready for a merge if this direction is appreciated: Shift the logic of global sidecars *configmap* to *helm* yaml.

File: pkg/operatormanager/operatormanager.go [hunk 1]
```diff
@@ -496,9 +502,10 @@ func (m *OperatorManager) createOrUpdateSidecarsConfig(ctx context.Context, name
 }
 
 func (m *OperatorManager) createOrUpdateSidecarsConfigMap(ctx context.Context, namespace string, c *v1.ConfigMap, objs []client.Object) ([]client.Object, error) {
+	sidecarsCMName := c.Data["per-namespace-sidecars-configmap-name"]
 
 	sccm := &v1.ConfigMap{}
-	if err := m.SetName(sccm, pg.SidecarsCMName); err != nil {
```
Comment (eberlep) [inline pkg/operatormanager/operatormanager.go:501]: Maybe move this `var` to operatormanager.

File: pkg/operatormanager/operatormanager.go [hunk 2]
```diff
@@ -570,10 +577,13 @@ func (m *OperatorManager) createOrUpdateExporterSidecarService(ctx context.Conte
...
-			Name:       "metrics",
-			Port:       int32(exporterServicePort),
-			Protocol:   v1.ProtocolTCP,
-			TargetPort: pg.ExporterSidecarPortName,
+			Name:     "metrics",
+			Port:     int32(exporterServicePort),
+			Protocol: v1.ProtocolTCP,
+			TargetPort: intstr.IntOrString{
+				Type:   intstr.String,
+				StrVal: "exporter",
```
Comment (eberlep) [inline pkg/operatormanager/operatormanager.go:585]: Hmm, maybe derive this value from the container definition? If someone changes the configmap an renames this port there, this service won't work anymore...

File: api/v1/postgres_types.go [hunk 3]
```diff
@@ -575,77 +507,42 @@ func init() {
...
-					CPU:    c.Data["postgres-fluentbit-requests-cpu"],
-					Memory: c.Data["postgres-fluentbit-requests-memory"],
-				},
-			},
-		},
+
+	// Deal with dynamically assigned name
+	for i := range sidecars {
+		for j := range sidecars[i].Env {
+			if sidecars[i].Env[j].Name == "DATA_SOURCE_PASS" {
```
Comment (eberlep) [inline api/v1/postgres_types.go:540]: this check should be more general: If there is a sidecars[i].Env[j].ValueFrom, we should then check if it is of the kind `SecretKeyRef`, and if so, we should proceed.

This implementation is just a workaround for one specific case.

**Review by eberlep — CHANGES_REQUESTED** [review]

File: charts/postgreslet/templates/configmap-sidecars.yaml [hunk 4]
```diff
@@ -1,20 +1,71 @@
...
-  postgres-exporter-requests-cpu: {{ .Values.sidecars.exporter.resources.requests.cpu | quote }}
-  postgres-exporter-requests-memory: {{ .Values.sidecars.exporter.resources.requests.memory | quote }}
-  postgres-exporter-limits-cpu: {{ .Values.sidecars.exporter.resources.limits.cpu | quote }}
-  postgres-exporter-limits-memory: {{ .Values.sidecars.exporter.resources.limits.memory | quote }}
-  postgres-fluentbit-image: {{ .Values.sidecars.fluentbit.image | quote }}
-  postgres-fluentbit-requests-cpu: {{ .Values.sidecars.fluentbit.resources.requests.cpu | quote }}
-  postgres-fluentbit-requests-memory: {{ .Values.sidecars.fluentbit.resources.requests.memory | quote }}
-  postgres-fluentbit-limits-cpu: {{ .Values.sidecars.fluentbit.resources.limits.cpu | quote }}
-  postgres-fluentbit-limits-memory: {{ .Values.sidecars.fluentbit.resources.limits.memory | quote }}
+  per-namespace-sidecars-configmap-name: {{ .Values.sidecars.perNamespaceConfigmapName | quote }}
```
[the remark this answers is not in the document]
Comment (LimKianAn) [reply charts/postgreslet/templates/configmap-sidecars.yaml:4]: We still need the variable [here](https://github.com/fi-ts/postgreslet/blob/215b172fc4ca0f4009207deaffb099dff9b9b2ef/charts/postgreslet/templates/configmap-sidecars.yaml#L54) and [here](https://github.com/fi-ts/postgreslet/blob/215b172fc4ca0f4009207deaffb099dff9b9b2ef/charts/postgreslet/templates/configmap-sidecars.yaml#L64). By the linkage of `per-namespace-sidecars-configmap-name` *go* code and *yaml* share a single source of truth.

File: pkg/operatormanager/operatormanager.go (hunk 1 above)
Comment (LimKianAn) [reply pkg/operatormanager/operatormanager.go:501]: As above, when *go* code and *yaml* share a variable, use the one in the *yaml*.

File: charts/postgreslet/templates/configmap-sidecars.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (LimKianAn) [reply charts/postgreslet/templates/configmap-sidecars.yaml:4]: This item is removed and instead derived dynamically in *go* code.

File: pkg/operatormanager/operatormanager.go (hunk 1 above)
Comment (LimKianAn) [reply pkg/operatormanager/operatormanager.go:501]: As above, the name is dynamically derived.

File: pkg/operatormanager/operatormanager.go (hunk 2 above)
Comment (LimKianAn) [reply pkg/operatormanager/operatormanager.go:585]: The configurable `.Values.sidecars.exporter.containerPort` in *values.yaml* is used now.

File: api/v1/postgres_types.go [hunk 5]
```diff
@@ -575,77 +507,42 @@ func init() {
...
-					Memory: c.Data["postgres-fluentbit-requests-memory"],
-				},
-			},
-		},
+
+	// Deal with dynamically assigned name
+	for i := range sidecars {
+		for j := range sidecars[i].Env {
+			if sidecars[i].Env[j].ValueFrom != nil && sidecars[i].Env[j].ValueFrom.SecretKeyRef != nil {
+				sidecars[i].Env[j].ValueFrom.SecretKeyRef.Name = "postgres." + p.ToPeripheralResourceName() + ".credentials"
```
Comment (eberlep) [inline api/v1/postgres_types.go:541]: This is still hardcoded for this usecase. I would prefer if this was completely generic. The name of the Secret should move back into the YAML/ConfigMap (the operator already supports the use of variables, in this case that string would look something like `postgres.{{ main.teamid }}-{{ main.name }}.credentials`). 

And while you're at it, you could also add `ConfigMapKeyRef`, `FieldRef` and `ResourceFieldRef` (https://pkg.go.dev/k8s.io/api@v0.20.4/core/v1?utm_source=gopls#EnvVarSource).

File: pkg/operatormanager/operatormanager.go [hunk 6]
```diff
@@ -683,3 +703,19 @@ func (m *OperatorManager) UpdateAllOperators(ctx context.Context) error {
 	m.log.Info("Done updating postgres operators in managed namespaces")
 	return nil
 }
+
+func perNamespaceSidecarsConfigMapName(cm *v1.ConfigMap) string {
```
Comment (eberlep) [inline pkg/operatormanager/operatormanager.go:707]: we should also talk about this, as already mentioned above. that looks too complicated to me?

File: pkg/operatormanager/operatormanager.go (hunk 1 above)
Comment (eberlep) [reply pkg/operatormanager/operatormanager.go:501]: now it's getting even more complicated? we create that configmap ourselves, and then derive it's name from another config we also created?

File: api/v1/postgres_types.go (hunk 5 above)
Comment (eberlep) [reply api/v1/postgres_types.go:541]: Now that we talk about it: remove that complete logic and only use `postgres.{{ main.teamid }}-{{ main.name }}.credentials` in the ConfigMap

Comment (eberlep) [reply api/v1/postgres_types.go:541]: OK, my bad: Try `{cluster}`

File: pkg/operatormanager/operatormanager.go (hunk 1 above)
Comment (LimKianAn) [reply pkg/operatormanager/operatormanager.go:501]: Final decision: Use hard-coded `const`.

File: api/v1/postgres_types.go (hunk 5 above)
Comment (LimKianAn) [reply api/v1/postgres_types.go:541]: No problem! The current setting in the `ConfigMap` is as follows: [`secret_name_template: '{username}.{cluster}.credentials'`](https://github.com/fi-ts/postgreslet/blob/28abc9171173c93f25151c1c7af593d2bfd45bc4/external/svc-postgres-operator.yaml#L45)

We can't use this feature in this case.

File: pkg/operatormanager/operatormanager.go (hunk 6 above)
Comment (LimKianAn) [reply pkg/operatormanager/operatormanager.go:707]: As above, use hard-coded `const` instead.

**Comment by LimKianAn** [comment]
Tests:
- 3 containers in pod are running. No errors in logs.
- `Service`postgres-exporter` is created. One can create an in-cluster curl pod  to pin the end-point and get html response.

Thus, merge this PR.

**Pull request merged**